### PR TITLE
Add Spanner Plugin

### PR DESCRIPTION
**Pull request by Ashi-Jain1**
Signed-off-by: Ashi Jain <Ashi.Jain1@ibm.com>

### CODE REVIEW

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+* The user can filter the logs using log name, resource, and severity.
+* In cases of multiple regions, use the same set of configurations for all regions. Use different configuration files for the input plug-in based on the region.
+
+##### Steps to Set destination(TOPIC & SUBSCRIPTION) permissions
+To set permissions for log sink to route to its destination, do the following:
+* Obtain the sink's writer identity—an email address—from the new sink.
+   1. Go to the Log Router page, and select menu  > View sink details. 
+   2. The writer identity appears in the Sink details panel.
+* If the user has owner access to the destination, 
+   1. add the sink's writer identity to topic and give it the Pub/Sub Publisher and Subscriber role.
```
Comment (AlexaNahum) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md:76]:
```suggestion
   1. add the sink's writer identity to the topic and give it the Pub/Sub Publisher and Subscriber role.
```

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+   2. add the sink's writer identity to subscription and give it the Pub/Sub Subscriber role.
```
Comment (AlexaNahum) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md:77]:
```suggestion
   2. add the sink's writer identity to the subscription and give it the Pub/Sub Subscriber role.
```

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+
+##### Steps to Create service account credentials
```
Comment (AlexaNahum) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md:80]:
```suggestion
##### Creating service account credentials
```

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+* Go to the Service accounts section of IAM & Admin console.
```
Comment (AlexaNahum) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md:81]:
```suggestion
* Go to the Service accounts section of the IAM & Admin console. 
```

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+   1. Go to the Log Router page, and select menu  > View sink details. 
+   2. The writer identity appears in the Sink details panel.
+* If the user has owner access to the destination, 
+   1. add the sink's writer identity to topic and give it the Pub/Sub Publisher and Subscriber role.
+   2. add the sink's writer identity to subscription and give it the Pub/Sub Subscriber role.
+
+
+##### Steps to Create service account credentials
+* Go to the Service accounts section of IAM & Admin console. 
+* Select project and click on Create Service Account.
```
Comment (AlexaNahum) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md:82]:
```suggestion
* Select project and click Create Service Account.
```

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+* Enter a Service account name, such as spanner-pubsub.
+* Click Create.
+* The service account needs access to a subscription that was created earlier. Use the Select a role drop-down menu to add the Pub/Sub Subscriber role. ​ 
+* Click Continue. The user does not need to grant users access to this service account.
```
Comment (AlexaNahum) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md:86]:
```suggestion
* Click Continue. You do not need to grant users access to this service account.
```

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+* Click the Create Key. The key is used by logstash input plug-in configuration file.
```
Comment (AlexaNahum) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md:87]:
```suggestion
* Click the Create Key. The key is used by the Logstash input plug-in configuration file.
```

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+* Select JSON and click Create. 
+
+##### Inclusion Filter
+Edit the Sink via *Logs Router > Sink Inclusion Filter*:
+##### Description
+The purpose of this inclusion filter is to include required and exclude unnecessary logs on the basis of some parameter from data_access and activity logs.
```
Comment (AlexaNahum) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md:93]:
```suggestion
The purpose of this inclusion filter is to include the required logs and exclude the unnecessary ones, on the basis of some parameter from data_access and activity logs.
```

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+
+```
+## 3. Supported audit logs
+* spanner-general.log - `INFO`, `DEFAULT`, `ALERT`,`NOTICE`,`DEBUG`,`WARNING`
+
+## 4. Configuring the Spanner filter in Guardium
+The Guardium universal connector is the Guardium entry point for native audit/data_access logs. The Guardium universal connector identifies and parses the received events, and converts them to a standard Guardium format. The output of the Guardium universal connector is forwarded to the Guardium sniffer on the collector, for policy and auditing enforcements. Configure Guardium to read the native audit/data_access logs by customizing the Spanner template.
+
+### Before you begin
+* You must have Log Full Details policy enabled on the collector. The detailed steps can be found in step 4 on [this page](https://www.ibm.com/docs/en/guardium/11.4?topic=dpi-installing-testing-filter-input-plug-in-staging-guardium-system).
```
Comment (AlexaNahum) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md:118]: per GRD-66763, change to: The detailed steps can be found in step 4 under section Installing and testing the filter or input plug-in on a staging Guardium system on [this page](https://github.com/IBM/universal-connectors/blob/main/docs/developing_plugins_gdp.md)

**Review by AlexaNahum — APPROVED** [review]

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java [hunk 10]
```diff
@@ -0,0 +1,523 @@
...
+		accessor.setDbProtocol(ApplicationConstants.DATA_PROTOCOL);
+		accessor.setDbProtocolVersion(ApplicationConstants.UNKOWN_STRING);
+
+		accessor.setOsUser(ApplicationConstants.UNKOWN_STRING);
+		accessor.setSourceProgram(ApplicationConstants.UNKOWN_STRING);
+
+		accessor.setClient_mac(ApplicationConstants.UNKOWN_STRING);
+		accessor.setServerDescription(ApplicationConstants.UNKOWN_STRING);
+
+		accessor.setDbUser(getAppUserName(protoPayloadJsonObject));
```
Comment (PratikshaSonawane) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java:117]: Please check if the assigned value is dbuser name or app user name? Please verify if both can be set to the same value.

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/test/java/com/ibm/guardium/spanner/SpannerFilterConnectorTest.java [hunk 11]
```diff
@@ -0,0 +1,745 @@
...
+		SpannerDBGuardiumFilter filter = getSpannerFilterConnector(gcpString2);
+		Event e = new org.logstash.Event();
+		TestMatchListener matchListener = new TestMatchListener();
+		e.setField("message", gcpString2);
+		Collection<co.elastic.logstash.api.Event> events = filter.filter(Collections.singletonList(e), matchListener);
+		assertNotNull(events);
+	}
```
Comment (PratikshaSonawane) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/src/test/java/com/ibm/guardium/spanner/SpannerFilterConnectorTest.java:83]: Here we have just added test cases where null checks are added over events, and no specific Fields are verified? Please add the correct test cases.

**Review by PratikshaSonawane — CHANGES_REQUESTED** [review]

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md [hunk 12]
```diff
@@ -0,0 +1,143 @@
+# Spanner-Guardium Logstash filter plug-in 
+
+This is a [Logstash](https://github.com/elastic/logstash) filter plug-in for the universal connector that is featured in IBM Security Guardium. It parses events and messages from Spanner audit/activity logs into a [Guardium record](https://github.com/IBM/universal-connectors/blob/main/common/src/main/java/com/ibm/guardium/universalconnector/commons/structures/Record.java) instance (which is a standard structure made out of several parts). The information is then pushed into Guardium. Guardium records include the accessor (the person who tried to access the data), the session, data, and exceptions. If there are no errors, the data contains details about the query "construct". The construct details the main action (verb) and collections (objects) involved.
```
[the remark this answers is not in the document]
Comment (Ashi-Jain1) [reply filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md:3]: Updated README.md as per the format.
Thanks

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/ApplicationConstants.java [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+
+SPDX-License-Identifier: Apache-2.0
+*/
+package com.ibm.guardium.spanner;
+
+import org.apache.commons.lang3.StringUtils;
+
+public class ApplicationConstants {
+
+	public static final String DATA_PROTOCOL = "Spanner(GCP)";
```
[the remark this answers is not in the document]
Comment (Ashi-Jain1) [reply filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/ApplicationConstants.java:12]: Since, this is a freetext field and there is no parser for spanner in Guardium, we kept it as Spanner (GCP). We went through the dev notes and implemented in suggested manner. Even if any change is required, we will do it post-release.

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java [hunk 14]
```diff
@@ -0,0 +1,523 @@
...
+		record.setData(parseData(protoPayloadJsonObject));
+
+		record.setTime(parseTime(inputJson));
+
+		return record;
+	}
+
+	private static String getSessionId(JsonObject protoPayloadJsonObject, JsonObject resourceJsonObject) {
+		Integer sessionId = (getCallerIp(protoPayloadJsonObject) + getAppUserName(protoPayloadJsonObject)
+				+ getInstanceNameAndProjectId(resourceJsonObject, ApplicationConstants.COLON)
```
[the remark this answers is not in the document]
Comment (Ashi-Jain1) [reply filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java:82]: We looked at the code, as it is not impacting the sessionId. If required, we will do it post-release.

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java (hunk 10 above)
Comment (Ashi-Jain1) [reply filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java:117]: In GCP platform we are using principalEmail to connect as well as same id is being used for Spanner DB Connection thats why we are keeping same email in both the fields.

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java [hunk 15]
```diff
@@ -0,0 +1,523 @@
...
+	 * @param JsonObject protoPayloadJsonObject
+	 * @return String value appUserName
+	 * 
+	 */
+	private static String getAppUserName(JsonObject protoPayloadJsonObject) {
+		String appUserName = StringUtils.EMPTY;
+		if (!protoPayloadJsonObject.has(ApplicationConstants.AUTHENTICATION_INFO)) {
+			return appUserName;
+		}
+		JsonObject authenticationJSON = protoPayloadJsonObject.get(ApplicationConstants.AUTHENTICATION_INFO)
```
[the remark this answers is not in the document]
Comment (Ashi-Jain1) [reply filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java:468]: We looked at the code; it is written in such a way that it will not throw a NullPointerException. In line 465, it is checking whether protoPayload has the Authentication_Info attribute or not.

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/test/java/com/ibm/guardium/spanner/SpannerFilterConnectorTest.java (hunk 11 above)
Comment (Ashi-Jain1) [reply filter-plugin/logstash-filter-pubsub-spanner-guardium/src/test/java/com/ibm/guardium/spanner/SpannerFilterConnectorTest.java:83]: We will do the changes post-release.

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+4. Click Upload File and select the key.json file. After it is uploaded, click OK.
+5. Click the Plus sign to open the Connector Configuration dialog box.
+6. Type a name in the Connector name field.
+7. Update the input section to add the details from [spanner_with_pubsub.conf](spanner_with_pubsub.conf) file's input part, omitting the keyword "input{" at the beginning and its corresponding "}" at the end.
+8. Update the filter section to add the details from [spanner_with_pubsub.conf](spanner_with_pubsub.conf) file's filter part, omitting the keyword "filter{" at the beginning and its corresponding "}" at the end.
+9.  The 'type' fields should match in input and filter configuration sections. This field should be unique for every individual connector added.
+10. Click Save. Guardium validates the new connector, and enables the universal connector if it was disabled. After it is validated, it appears in the Configure Universal Connector page.
+11. After installing the plugins offline package and once the configuration is uploaded and saved to your Guardium machine, restart the universal connector using the Disable/Enable button.
+
+## 5. Limitations
```
Comment (PratikshaSonawane) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md:132]: Please move this section just above "Configuring the Spanner filter in Guardium"

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java (hunk 10 above)
Comment (PratikshaSonawane) [reply filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java:117]: For SpannerDB connection, I agree we use priincipalEmail. But where else do you want to say by GCP platform to connect?
What I understand is appUsername is the Username with which the user logged in to the App used to execute the queries. @Tal-Daniel Please correct me if I am wrong.
Same issue was raised for couch-db-on-prem, where it was fixed then, please check
Thank you

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (PratikshaSonawane) [reply filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java:82]: @Tal-Daniel  Please suggest

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (PratikshaSonawane) [reply filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java:468]: Ok thanks

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/test/java/com/ibm/guardium/spanner/SpannerFilterConnectorTest.java (hunk 11 above)
Comment (PratikshaSonawane) [reply filter-plugin/logstash-filter-pubsub-spanner-guardium/src/test/java/com/ibm/guardium/spanner/SpannerFilterConnectorTest.java:83]: @Tal-Daniel  Please suggest

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (PratikshaSonawane) [reply filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md:3]: Limitations section is placed at the bottom, please move it as suggested. Thanks

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/ApplicationConstants.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (Tal-Daniel) [reply filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/ApplicationConstants.java:12]: Data protocol have some freedom for developers' decision, as far I know. 
@Ashi-Jain1, could you compare with other filter plug-ins that work on Google cloud? If you leave GCP in the end, then add a space "Spanner (GCP)". 

Also, if Spanner solely exist on Google cloud, it's needless tot mention that in parenthesis, TMHO. 

This constant change can be done in this PR, especially if no manipulation or conditional depend on it.

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (Tal-Daniel) [reply filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java:82]: I see where this is coming from: That's how you represent the DB name (service name).

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java (hunk 10 above)
Comment (Ashi-Jain1) [reply filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java:117]: Hi Pratiskha,
Here as well we use key file which has principalEmail where the same email id is getting used.
But still if change is required, we will take it as post release.

Thanks

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (Ashi-Jain1) [reply filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md:3]: Hello @PratikshaSonawane ,

We have followed the guidelines provided in the template.
As per the template, limitations should be at the end.
PFA 
[guidelines.txt](https://github.com/IBM/universal-connectors/files/10453852/guidelines.txt)

[[UCX-146] [DocumentDB UC]:Readme file changes required - IBM Security](https://jira-ds.rtp.raleigh.ibm.com:8443/browse/UCX-146)

Thanks

**Review by Tal-Daniel — CHANGES_REQUESTED** [review]
Is the org/perboiled code originally developed by you? If not, we may need to add a NOTICE file, and keep original notice in files.

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/test/java/com/ibm/guardium/spanner/SpannerFilterConnectorTest.java (hunk 11 above)
Comment (Tal-Daniel) [reply filter-plugin/logstash-filter-pubsub-spanner-guardium/src/test/java/com/ibm/guardium/spanner/SpannerFilterConnectorTest.java:83]: Ashi, please add tests following this PR, as Pratiksha implies, that validate the Record contents, not only not null. 
It's recommended to add tests as soon as possible in the process.

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/ApplicationConstants.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (Tal-Daniel) [reply filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/ApplicationConstants.java]: Thanks Ashi. Resolved as changed to "Spanner. no need for retesting, as this is a language that Guardiam does not parse, but rather you supply the parser.

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java (hunk 10 above)
Comment (Tal-Daniel) [reply filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java:117]: Correct, @PratikshaSonawane. If no app username, please leave it empty, Ashi, as this is an optional field (according to [Record.java](https://github.com/IBM/universal-connectors/blob/main/common/src/main/java/com/ibm/guardium/universalconnector/commons/structures/Record.java#L53-L56))

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (Tal-Daniel) [reply filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java:82]: This is not interfering with the session hashing.

**Comment by Tal-Daniel** [comment]
> Is the org/perboiled code originally developed by you? If not, we may need to add a NOTICE file, and keep original notice in files.

Ashi, a kind reminder.

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java (hunk 10 above)
Comment (Ashi-Jain1) [reply filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java:117]: Done the changes, changed the value of AppUserName to empty.

Thanks

**Comment by Ashi-Jain1** [comment]
> > Is the org/perboiled code originally developed by you? If not, we may need to add a NOTICE file, and keep original notice in files.
> 
> Ashi, a kind reminder.

Hi Tal,

Yes, this code is developed by HCL, confusion may have arisen due to the org package name. Will modify the package name post-release.
We have also sent the same over mail 1/23/2023. (Spanner PR Comments)

**Review by PratikshaSonawane — APPROVED** [review]

**Review by Tal-Daniel — APPROVED** [review]
Approved. Readme and unit tests comments will be resolved post-release.
